src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,779 @@
+            _ => None,
+        }
+    }
+
+    /// Gets the column offset for `ColumnRef` variant.
+    #[inline]
+    pub fn column_ref_offset(&self) -> Option<usize> {
+        match self {
+            RpnExpressionNode::ColumnRef { ref offset, .. } => Some(*offset),
+            _ => None,
+        }
+    }
+}
+
+/// An RPN expression node list which represents an expression in Reverse Polish notation.
+#[derive(Debug)]
+pub struct RpnExpressionNodeVec(Vec<RpnExpressionNode>);
+
+impl std::ops::Deref for RpnExpressionNodeVec {
+    type Target = Vec<RpnExpressionNode>;
+
+    fn deref(&self) -> &Vec<RpnExpressionNode> {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for RpnExpressionNodeVec {
+    fn deref_mut(&mut self) -> &mut Vec<RpnExpressionNode> {
+        &mut self.0
+    }
+}
+
+impl RpnExpressionNodeVec {
+    /// Builds the RPN expression node list from an expression definition tree.
+    // TODO: Deprecate it in Coprocessor V2 DAG interface.
+    pub fn build_from_expr_tree(tree_node: Expr, time_zone: Tz) -> Result<Self> {
+        let mut expr_nodes = Vec::new();
+        Self::append_rpn_nodes_recursively(
+            tree_node,
+            time_zone,
+            &mut expr_nodes,
+            super::map_pb_sig_to_rpn_func,
+        )?;
+        Ok(RpnExpressionNodeVec(expr_nodes))
+    }
+
+    /// Transforms eval tree nodes into RPN nodes.
+    ///
+    /// Suppose that we have a function call:
+    ///
+    /// ```ignore
+    /// A(B, C(E, F, G), D)
+    /// ```
+    ///
+    /// The eval tree looks like:
+    ///
+    /// ```ignore
+    ///           +---+
+    ///           | A |
+    ///           +---+
+    ///             |
+    ///   +-------------------+
+    ///   |         |         |
+    /// +---+     +---+     +---+
+    /// | B |     | C |     | D |
+    /// +---+     +---+     +---+
+    ///             |
+    ///      +-------------+
+    ///      |      |      |
+    ///    +---+  +---+  +---+
+    ///    | E |  | F |  | G |
+    ///    +---+  +---+  +---+
+    /// ```
+    ///
+    /// We need to transform the tree into RPN nodes:
+    ///
+    /// ```ignore
+    /// B E F G C D A
+    /// ```
+    ///
+    /// The transform process is very much like a post-order traversal. This function does it
+    /// recursively.
+    fn append_rpn_nodes_recursively<F>(
+        mut tree_node: Expr,
+        time_zone: Tz,
+        rpn_nodes: &mut Vec<RpnExpressionNode>,
+        fn_mapper: F,
+    ) -> Result<()>
+    where
+        F: Fn(tipb::expression::ScalarFuncSig) -> Result<Box<dyn RpnFunction>> + Copy,
+    {
+        // TODO: We should check whether node types match the function signature. Otherwise there
+        // will be panics when the expression is evaluated.
+
+        use crate::coprocessor::codec::mysql::{Decimal, Duration, Json, Time, MAX_FSP};
+        use crate::util::codec::number;
+        use std::convert::{TryFrom, TryInto};
+
+        #[inline]
+        fn handle_node_null(
```

Comment:
Could we move these functions outside this function? It's hard to review.

Reply:
I'm very confused about the code style in rust, Is it a good style to define lots of functions inside the function which would use it? @BusyJay @brson

Reply:
I can't see moving functions to outside (1 less level of ident) will make any differences 😆 

Additionally, to move to outside, these function will be more complex. More specifically:
- for each function it needs to **NOT** accept EvalType. Instead for each function it needs to parse EvalType from `expr.get_field_type().tp()`. This will introduce a lot of code duplications.
- the function name will be more specific, to indicate that it only works for the process or transforming expr tree into RPN, which means, it will be much longer.

Reply:
@AndreMouche I don't recall ever seeing any discussion about whether it's good or bad to define functions inside functions like this, but I can offer my opinion.

I do it, but don't have any particular rules I follow about where it's appropriate. Mostly I do it for small helper functions, defining them close to where they are called. Sometimes I'll define a few of them either at the top or bottom of the outer function.

All else considered, I might not personally write this function nested as it is, because it extends several screens, making it easy to lose track of which outer functions the inner functions are in service of - pretty much the same principle as keeping functions small generally for readability. (On the other hand though, having them defined internally also makes it explicit _exactly_ which function they are in service of).

One thing that could be done differently here that could make it more readable is to place the outer function's content _before_ the inner function's rather than after. That way when you start reading the function, you are reading the substance of the function, and whatever comes after, regardless of the indent level is something else. In other words, if I want to read this function, I want to look for its header, then read the body; not look for the header, scan down several screens while matching brackets mentally, then read the body.

The only slightly unfortunate thing about placing the function body before the inner definitions instead of after is that the function can't end with an expression, but rather needs an explicit `return` statement.

@overvenus makes 2 observations about why it would be undesirable to move these outside:

>     * for each function it needs to **NOT** accept EvalType. Instead for each function it needs to parse EvalType from `expr.get_field_type().tp()`. This will introduce a lot of code duplications.

I don't follow this. Since these are all regular functions, with no context, and not methods, I'd think they could be defined exactly the same inside or out.

> the function name will be more specific, to indicate that it only works for the process or transforming expr tree into RPN, which means, it will be much longer.

This seems reasonable. If it seems easy to misuse the functions, then defining them inside definitely prevents that. An alternative would be to define a module to contain all the functions.

In my opinion in this case ... I don't have a strong opinion, but I am inclined to move the function body to the top of the function and evaluating whether that is more readable.

I'm curious though what others think. cc @nrc

Reply:
@brson I didn't make it clear. What I mean is that, the function accepts a parameter called `tree_node: Expr` and a parameter called `field_type: FieldType`. The `field_type` is calculated **from** `tree_node`. The implementation of the function relies on both parameters to be matching, i.e. the given `field_type` is exactly the one that is calculated from the given `tree_node`. Otherwise the result may be wrong or unexpected.

Simply speaking, here all these small functions have prerequisites: `field_type` and `tree_node` matches. Here these functions are inside the function that provides both, so there is no way to supply unmatched two parameters. However, when they are moved to outside, we need to forbid this.

There are some solutions:

1. Current one: Keep the function inside.

2. Explicitly document the prerequisite of each function by using doc comments, that `field_type` and `tree_node` must match otherwise the result can be incorrect. This heavily inflates the code.

3. Only accept the `tree_node` for each function. This requires each function calculates `field_type` from `tree_node` by their own, also inflates the code pretty much.

About the naming issue that I mentioned above, I also need to supply a background knowledge that this PR is the first PR that adds some code to the module. The module is used to perform vectorized evaluation, while this PR only includes the creation part and the evaluation part (which is the main functionality of the module) is just not included. Thus, making function names longer to clearly show that these functions are used for the creation instead of evaluating is very necessary.

Reply:
@brson By the way actually I don't like these small functions. They are different implementations of each `match` arm and they are never called in two places. Do you think it's better to just make them not functions. Instead, just fine to be inlined in each `match` arm?

Reply:
For me I'd factor the common parts as a macro 🙃

Reply:
My thoughts are that given this function has its own imports, invariants, and multiple sub-functions, it would be better as a module rather than a function (small modules are good!). I didn't read the code, but from these comments it sounds like perhaps field_type and tree_node should be fields in a small struct and that the inline functions should be methods on that struct?

Reply:
Thanks for the detailed explanation @breeswish!